### PR TITLE
Enforce tenancy (second try)

### PR DESCRIPTION
**Pull request by gtanzillo**
Enforce tenancy by default. Tenancy can be bypassed by setting the environment variable `BYPASS_TENANCY` to "true"

This PR is a replacement of the [original PR](https://github.com/ManageIQ/topological_inventory-api/pull/104) which got out of sync with master due to moving and reworking on many tests. The main changes and tests that directly test tenancy have been pulled in from that PR. 

However, updates to the many existing tests to make them pass with enforcement on by default have been left behind because the in order to get something that can be merged soon.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
 
   private
 
+  set_current_tenant_through_filter
+  before_action :set_the_current_tenant
+  after_action  :unset_the_current_tenant
+
+  def set_the_current_tenant
+    return unless Tenant.tenancy_enabled?
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: If tenancy is disabled, then doesn't the sources_controller end up creating a tenant anyway since it executes before this line?

File: app/controllers/api/v0/sources_controller.rb [hunk 2]
```diff
@@ -6,6 +6,13 @@ class SourcesController < ApplicationController
       include Api::V0::Mixins::ShowMixin
       include Api::V0::Mixins::UpdateMixin
 
+      def set_the_current_tenant
+        # Auto-create tenant if creating a new source and tenant does not exist
+        Tenant.find_or_create_by(:external_tenant => user_identity) if action_name == "create"
```
Comment (Fryguy) [inline app/controllers/api/v0/sources_controller.rb]: why is user_identity stored in the external_tenant?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
...
+
+    tenant = Tenant.find_by(:external_tenant => user_identity)
+    if tenant
+      set_current_tenant(tenant)
+    else
+      error_document = {
+          "errors" => [
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: 2 space indent

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
...
   private
 
+  set_current_tenant_through_filter
+  before_action :set_the_current_tenant
+  after_action  :unset_the_current_tenant
+
+  def set_the_current_tenant
+    return unless Tenant.tenancy_enabled?
+
+    tenant = Tenant.find_by(:external_tenant => user_identity)
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: Can we just let this blow up naturally with an ActiveRecord::RecordNotFound and go through the normal path instead of building our own error document?

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
 
   private
 
+  set_current_tenant_through_filter
+  before_action :set_the_current_tenant
+  after_action  :unset_the_current_tenant
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: This might be better designed as an around_action since they are a joined pair.

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
...
+      }
+      render :json => error_document, :status => :unauthorized
+    end
+  end
+
+  def unset_the_current_tenant
+    set_current_tenant(nil)
+  end
+
+  def user_identity
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: Recommend renaming since it's not the user_identity, but the user's tenant.

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
...
+    ManageIQ::API::Common::Request.with_request(request) do |current|
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: This should probably be around the entire request with a before_action (though if that's breaking a lot of things we can do in a followup)

File: app/controllers/api/v0/sources_controller.rb (hunk 2 above)
Comment (gtanzillo) [reply app/controllers/api/v0/sources_controller.rb]: It's the account number. I'll rename that method

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -22,6 +22,39 @@ class ApplicationController < ActionController::API
...
+      current.user.tenant if current.headers.key?(ManageIQ::API::Common::Request::IDENTITY_KEY)
```
Comment (Fryguy) [inline app/controllers/application_controller.rb]: checking the key this way may not be case-insensitive, so this might not be a accurate.  Perhaps just get the current.user.tenant and catch the exception instead?  It should be set, so exception handling feels like the right path.  Additionally, if there is no tenant, what happens?  This code doesn't raise, so it just moves on?

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (gtanzillo) [reply app/controllers/application_controller.rb]: I tried to do it that way but it seemed to bypass Brandon's code that checks parameters so I backed it out. Let me try again and maybe we can look at it together

File: spec/support/tenant_identity.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+      included do
+        let!(:tenant)           { Tenant.create!(:external_tenant => external_tenant) }
+        let!(:external_tenant)  { rand(1000).to_s }
+        let!(:unknown_tenant)   { rand(1000).to_s }
+        let!(:identity)         { Base64.encode64({'identity' => { 'account_number' => external_tenant}}.to_json) }
+        let!(:unknown_identity) { Base64.encode64({'identity' => { 'account_number' => unknown_tenant}}.to_json) }
+      end
+    end
+  end
+end
```
Comment (Fryguy) [inline spec/support/tenant_identity.rb]: missing CR at end of file.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (gtanzillo) [reply app/controllers/application_controller.rb]: Yes, but I didn't think it would hurt to let it do that

File: spec/support/tenant_identity.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  module Support
+    module TenantIdentity
+      extend ActiveSupport::Concern
+
+      included do
+        let!(:tenant)           { Tenant.create!(:external_tenant => external_tenant) }
+        let!(:external_tenant)  { rand(1000).to_s }
+        let!(:unknown_tenant)   { rand(1000).to_s }
+        let!(:identity)         { Base64.encode64({'identity' => { 'account_number' => external_tenant}}.to_json) }
+        let!(:unknown_identity) { Base64.encode64({'identity' => { 'account_number' => unknown_tenant}}.to_json) }
```
Comment (Fryguy) [inline spec/support/tenant_identity.rb]: Feels like we should have helper methods in ManageIQ::API::Common for this...future follow up. cc @syncrou

**Comment by Fryguy** [comment]
Looking good...got a number of comments above.

File: spec/support/tenant_identity.rb (hunk 10 above)
Comment (syncrou) [reply spec/support/tenant_identity.rb]: @Fryguy, @gtanzillo  - Yes.  We actually already do.  https://github.com/ManageIQ/manageiq-api-common/blob/master/spec/support/user_header_spec_helper.rb

Comment (syncrou) [reply spec/support/tenant_identity.rb]: There are a number of examples in the manageiq-common and the catalog where a new key/value is applied and re Base64 encoded as well.

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (gtanzillo) [reply app/controllers/application_controller.rb]: I think I can get this to work. Looking at it now

File: spec/controllers/api/v0x0/application_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,62 @@
+RSpec.describe ApplicationController, :type => :request do
```
Comment (agrare) [inline spec/controllers/api/v0x0/application_controller_spec.rb]: `spec/controllers/api/v0x0/application_controller_spec.rb`

@gtanzillo should this be in `spec/controllers/application_controller_spec.rb` instead of under `api/v0x0` ?

Comment (gtanzillo) [reply spec/controllers/api/v0x0/application_controller_spec.rb]: Yes, I'll move it

File: app/controllers/api/v0/sources_controller.rb [hunk 12]
```diff
@@ -6,6 +6,17 @@ class SourcesController < ApplicationController
       include Api::V0::Mixins::ShowMixin
       include Api::V0::Mixins::UpdateMixin
 
+      def with_current_request
+        ManageIQ::API::Common::Request.with_request(request) do |current|
+          if Tenant.tenancy_enabled?
+            # Auto-create tenant if creating a new source and tenant does not exist
+            Tenant.find_or_create_by(:external_tenant => current.user.tenant) if action_name == "create" rescue KeyError
```
Comment (agrare) [inline app/controllers/api/v0/sources_controller.rb]: Do you need to set the current tenant here after creating it?

Comment (gtanzillo) [reply app/controllers/api/v0/sources_controller.rb]: Should be handled by `super`. But, in looking at that, I saw the the test was not quite correct.
